packages/shim/src/lib/shim.ts

```diff
@@ -0,0 +1,340 @@
+    return global.__SENTRY__.processStack;
+  }
+
+  /**
+   * TODO
+   */
+  private getProcessStackTop(): ScopeLayer {
+    const stack = this.getProcessStack();
+    if (stack.length === 0) {
+      stack.push({
+        scope: {},
+        type: 'process',
+      });
+    }
+    return stack[stack.length - 1];
+  }
+
+  /**
+   * TODO
+   */
+  private getDomainStack(): ScopeLayer[] | undefined {
+    if (!shimDomain.active) {
+      return undefined;
+    }
+    let sentry = shimDomain.active.__SENTRY__;
+    if (!sentry) {
+      shimDomain.active.__SENTRY__ = sentry = { domainStack: [] };
+    }
+    return sentry.domainStack;
+  }
+
+  /**
+   * TODO
+   */
+  private getDomainStackTop(): ScopeLayer | undefined {
+    const stack = this.getDomainStack();
+    if (stack === undefined) {
+      return undefined;
+    }
+    if (stack.length === 0) {
+      stack.push({
+        client: getCurrentClient(),
+        scope: this.getInitialScope(),
+        type: 'domain',
+      });
+    }
+    return stack[stack.length - 1];
+  }
+
+  /**
+   * TODO
+   */
+  private getInitialScope(): any {
+    let initalScope = {};
+    try {
+      initalScope = getCurrentClient() && getCurrentClient().getInitialScope();
+    } catch {
+      // we do nothing
+    }
+    return initalScope;
+  }
+}
+
+/**
+ * TODO
+ */
+function _getLatestShim(): Shim {
+  if (
+    global.__SENTRY__.shim === undefined ||
+    global.__SENTRY__.shim.isOlderThan(API_VERSION)
+  ) {
+    global.__SENTRY__.shim = new Shim();
+  }
+  return global.__SENTRY__.shim;
+}
+
+/**
+ * TODO
+ * @param promise
+ */
+function forget(promise?: any): void {
+  if (promise && typeof promise.catch === 'function') {
+    promise.catch((e: any) => {
+      console.error(e);
+    });
+  }
+}
+
+/**
+ * TODO
+ */
+export function pushScope(client?: any): void {
+  _getLatestShim().pushScope(client);
+}
+
+/**
+ * TODO
+ */
+export function popScope(): void {
+  _getLatestShim().popScope();
+}
+
+/**
+ * TODO
+ */
+export function withScope(client: any, callback: () => void): void;
+export function withScope(callback: () => void, client?: any): void;
+export function withScope(arg1: any, arg2: any): void {
+  _getLatestShim().withScope(arg1, arg2);
+}
+
+/**
+ * TODO
+ */
+export function clearScope(): void {
+  _getLatestShim().clearScope();
+}
+
+/**
+ * TODO
+ */
+export function getCurrentClient(): any | undefined {
+  return _getLatestShim().getStackTop().client;
+}
+
+/**
+ * TODO
+ */
+export function bindClient(client: any): void {
+  const top = _getLatestShim().getStackTop();
+  top.client = client;
+  top.scope = client.getInitialScope();
+}
+
+// api
+/**
+ * Captures an exception evebt and sends it to Sentry.
+ *
+ * @param exception An exception-like object.
+ * TODO
+ * @returns A Promise that resolves when the exception has been sent.
+ */
+export function captureException(exception: any): void {
+  const top = _getLatestShim().getStackTop();
+  if (top.client) {
```

Comment:
This shouuuldn't ever break as `processStack` is initialized with `[]`, but I'd check `top && top.client` anyway, just for the sake of safety. Same applies to all the calls below.